File: cpp/127.cpp
class Solution {
private:
    void addNextWord(string &word, unordered_set<string>& wordList, queue<string>&visit){
        wordList.erase(word);
        for(int p=0; p<word.size(); p++){
            char ch = word[p];
            for(int i=0; i<26; i++){
                word[p] = 'a' + i;
                if(wordList.find(word)!=wordList.end()){
                    visit.push(word);
                    wordList.erase(word);
                }
            }
            word[p] = ch;
        }
    }
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordList) {
        if(beginWord == endWord) return 0;
        wordList.insert(endWord);
        queue<string> visit;
        addNextWord(beginWord, wordList, visit);
        int step = 2;
        while(!visit.empty()){
            int num = visit.size();
            for(int i=0; i<num; i++){
                string word = visit.front();
                if(word==endWord){
                    return step;
                }
                visit.pop();
                addNextWord(word, wordList, visit);
            }
            step++;
        }
        return 0;
    }
};

File: golang/127/127.go
package main

type word struct {
	str   string
	level int
}

func ladderLength(beginWord string, endWord string, wordList []string) int {
	queue := []word{
		{
			str:   beginWord,
			level: 1,
		},
	}
	pos := 0
	used := make([]bool, len(wordList))
	for pos < len(queue) {
		level := queue[pos].level + 1
		for i, v := range wordList {
			if used[i] {
				continue
			}

			if canChange(queue[pos].str, v) {
				if v == endWord {
					return level
				}
				queue = append(queue, word{
					str:   v,
					level: level,
				})
				used[i] = true
			}
		}
		pos++
	}

	return 0
}

func canChange(a, b string) bool {
	if len(a) != len(b) {
		return false
	}

	diff := 0
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			diff++
			if diff > 1 {
				return false
			}
		}
	}
	return true
}
